File: go/chars.go
// Package leanprompt is a zero-dependency prompt-compression pipeline for LLM
// applications: a deterministic extractive compressor (Extract) behind a
// classifier-gated middleware, implementing docs/parity-spec.md and
// byte-compatible with the TypeScript reference (parity/*.json).
package leanprompt

// Pinned character classes (parity-spec §2). Host Unicode tables
// (unicode.IsLetter etc.) vary by Unicode version and are never used.

func isSpaceChar(cp rune) bool {
	switch {
	case cp == 0x20, cp >= 0x09 && cp <= 0x0d, cp == 0x85, cp == 0xa0:
		return true
	case cp >= 0x2000 && cp <= 0x200b, cp == 0x2028, cp == 0x2029,
		cp == 0x202f, cp == 0x205f, cp == 0x3000, cp == 0xfeff:
		return true
	}
	return false
}

func isPunctChar(cp rune) bool {
	switch {
	case cp >= 0x2010 && cp <= 0x2027, cp >= 0x2030 && cp <= 0x205e,
		cp >= 0x3001 && cp <= 0x303f, cp >= 0xfe50 && cp <= 0xfe6f,
		cp >= 0xff01 && cp <= 0xff0f, cp >= 0xff1a && cp <= 0xff20,
		cp >= 0xff3b && cp <= 0xff40, cp >= 0xff5b && cp <= 0xff65:
		return true
	}
	return false
}

// isWordChar: ASCII alphanumerics; any non-ASCII code point that is neither
// pinned whitespace nor pinned punctuation.
func isWordChar(cp rune) bool {
	if cp < 0x80 {
		return (cp >= '0' && cp <= '9') || (cp >= 'A' && cp <= 'Z') || (cp >= 'a' && cp <= 'z')
	}
	return !isSpaceChar(cp) && !isPunctChar(cp)
}

// asciiLower maps A-Z to a-z only; every other code point is unchanged.
func asciiLower(text string) string {
	out := make([]rune, 0, len(text))
	for _, c := range text {
		if c >= 'A' && c <= 'Z' {
			c += 32
		}
		out = append(out, c)
	}
	return string(out)
}

File: rust/src/chars.rs
//! Pinned character classes (parity-spec §2).
//!
//! Host Unicode tables (`char::is_alphabetic` etc.) vary by Unicode version
//! and must not be used anywhere in the pipeline.

/// Pinned whitespace rule (code point).
pub fn is_space_char(cp: u32) -> bool {
    matches!(cp,
        0x20 | 0x09..=0x0d | 0x85 | 0xa0
        | 0x2000..=0x200b | 0x2028 | 0x2029 | 0x202f | 0x205f
        | 0x3000 | 0xfeff)
}

fn is_punct_char(cp: u32) -> bool {
    matches!(cp,
        0x2010..=0x2027 | 0x2030..=0x205e | 0x3001..=0x303f
        | 0xfe50..=0xfe6f | 0xff01..=0xff0f | 0xff1a..=0xff20
        | 0xff3b..=0xff40 | 0xff5b..=0xff65)
}

/// Pinned word-character rule (code point): ASCII alphanumerics; any
/// non-ASCII code point that is neither pinned whitespace nor pinned
/// punctuation.
pub fn is_word_char(cp: u32) -> bool {
    if cp < 0x80 {
        return (0x30..=0x39).contains(&cp)
            || (0x41..=0x5a).contains(&cp)
            || (0x61..=0x7a).contains(&cp);
    }
    !is_space_char(cp) && !is_punct_char(cp)
}

/// ASCII-only lowercasing: `A-Z` → `a-z`, everything else unchanged.
pub fn ascii_lower(text: &str) -> String {
    text.chars()
        .map(|c| if c.is_ascii_uppercase() { c.to_ascii_lowercase() } else { c })
        .collect()
}
